File: Evaluate Reverse Polish Notation.java
// Time Complexity : O(n), where n is the number of tokens in the input array.
// Space Complexity : O(n), in the worst case, all tokens are numbers and pushed onto the stack.
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No major issues. Just had to ensure operator order (for '-' and '/') is correct.

// Your code here along with comments explaining your approach:

class Solution {
    public int evalRPN(String[] tokens) {
        // Stack to hold operands
        Stack<Integer> st = new Stack<>();
        
        // Iterate through all tokens
        for (int i = 0; i < tokens.length; ++i) {
            String x = tokens[i];
            
            // If current token is an operator
            if (x.equals("+")) {
                int a = st.pop(); // second operand
                int b = st.pop(); // first operand
                st.push(b + a);
            } else if (x.equals("-")) {
                int a = st.pop();
                int b = st.pop();
                st.push(b - a); // order matters: b - a
            } else if (x.equals("*")) {
                int a = st.pop();
                int b = st.pop();
                st.push(b * a);
            } else if (x.equals("/")) {
                int a = st.pop();
                int b = st.pop();
                st.push(b / a); // order matters: b / a
            } else {
                // It's a number, push to stack
                int val = Integer.parseInt(x);
                st.push(val);
            }
        }

        // Final result is the only item left on the stack
        return st.peek();
    }
}

File: Evaluate Reverse Polish Notation.cpp
// Time Complexity : O(n), where n is the number of tokens
// Space Complexity : O(n), for storing numbers in the stack
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : 
// - Needed to handle negative numbers carefully (isdigit(x[0]) isn't enough for negatives)
// - Ensured correct operand order in subtraction and division (b - a, b / a)

// Your code here along with comments explaining your approach:

class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        int n = tokens.size();
        stack<int> st;

        // Loop through all tokens
        for (int i = 0; i < n; ++i) {
            string x = tokens[i];

            // Check if the token is an operator
            // Note: need to handle negative numbers, so check x.size() == 1 as well
            if (!isdigit(x[0]) && x.length() == 1) {
                // Pop top two numbers from stack
                int a = st.top(); st.pop();
                int b = st.top(); st.pop();
                // Perform the operation and push result back
                st.push(performOperation(x, a, b));
            } else {
                // Convert operand string to integer and push to stack
                st.push(stoi(x));
            }
        }

        // Final result is on top of the stack
        return st.top();
    }

    // Helper function to perform the given operation
    int performOperation(string x, int a, int b) {
        switch (x[0]) {
        case '+': return a + b;
        case '-': return b - a;
        case '*': return a * b;
        case '/': return b / a;
        }
        return 0; // should never reach here if input is valid
    }
};
